src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,49 +1,95 @@
 package core.basesyntax;
 
-public class ArrayList<T> implements List<T> {
+import java.util.Arrays;
+import java.util.NoSuchElementException;
 
+public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    
+    private T[] elements;
+    private int size;
+    
+    public ArrayList() {
+        elements = (T[]) new Object[DEFAULT_CAPACITY];
+        size = 0;
+    }
+    
     @Override
     public void add(T value) {
-
+        add(value, size);
     }
-
+    
     @Override
     public void add(T value, int index) {
-
+        checkAddIndex(index);
+        if (size == elements.length) {
+            grow();
+        }
+        System.arraycopy(elements, 0, elements, index, ++size - index);
+        elements[index] = value;
     }
-
+    
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
-
+    
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return elements[index];
     }
-
+    
     @Override
     public void set(T value, int index) {
-
+        checkIndex(index);
+        elements[index] = value;
     }
-
+    
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        T oldElement = get(index);
+        System.arraycopy(elements, index + 1, elements, index, size - index);
```

Comment:
```suggestion
        System.arraycopy(elements, index + 1, elements, index, ???);
```
the last part is not 100% correct, please reconsider it